### PR TITLE
mixin: ruler emote eval failing alert

### PR DESCRIPTION
**Pull request by ortuman**
The experimental `RulerRemoteEvaluationFailing` alert has proven to work correctly at the ops cluster. 

This PR moves the alert definition to upstream. 

#### Which issue(s) this PR fixes or relates to

Fixes N/A

#### Checklist

- [ ] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ### Grafana Mimir
 
 * [ENHANCEMENT] Distributor: Add age filter to forwarding functionality, to not forward samples which are older than defined duration. If such samples are not ingested, `cortex_discarded_samples_total{reason="forwarded-sample-too-old"}` is increased. #3049 #3133
+* [ENHANCEMENT] Alerts: added `RulerRemoteEvaluationFailing` alert, firing when communication between ruler and frontend fails in remote operational mode. #3177
```
Comment (osg-grafana) [inline CHANGELOG.md:8]:
```suggestion
* [ENHANCEMENT] Alerts: Added `RulerRemoteEvaluationFailing` alert, which fires when communication between the ruler and the frontend fails in remote operational mode. #3177
```

File: docs/sources/operators-guide/mimir-runbooks/_index.md [hunk 2]
```diff
@@ -346,6 +346,20 @@ How to **fix** it:
 - Increase the evaluation interval of the rule group. You can use the rate of missed evaluation to estimate how long the rule group evaluation actually takes.
 - Try splitting up the rule group into multiple rule groups. Rule groups are evaluated in parallel, so the same rules may still fit in the same resolution.
 
+### MimirRulerRemoteEvaluationFailing
+
+This alert fires when communication between `ruler` and `ruler-query-frontend` is failing to be established.
```
Comment (osg-grafana) [inline docs/sources/operators-guide/mimir-runbooks/_index.md:351]:
```suggestion
This alert fires when communication between the `ruler` and the `ruler-query-frontend` fails to become established.
```

File: docs/sources/operators-guide/mimir-runbooks/_index.md [hunk 3]
```diff
@@ -346,6 +346,20 @@ How to **fix** it:
...
+
+The `ruler-query-frontend` component is exclusively used by the `ruler` to evaluate rule expressions when running in remote operational mode. If communication between these two components breaks, gaps are expected to appear in the case of recording rules or alerting rules will not fire when they should.
```
Comment (osg-grafana) [inline docs/sources/operators-guide/mimir-runbooks/_index.md:353]:
```suggestion
The `ruler-query-frontend` component is exclusively used by the `ruler` to evaluate rule expressions while running in remote operational mode. If communication between these two components breaks, gaps appear, and recording rules or alerting rules do not fire when they should.
```
Please vet my changes for technical accuracy.

File: docs/sources/operators-guide/mimir-runbooks/_index.md [hunk 4]
```diff
@@ -346,6 +346,20 @@ How to **fix** it:
...
+
+How to **investigate**:
+
+- Check the `Mimir / Remote ruler reads` dashboard to see if the issue is caused by failures or high latency
+  - **Failures**
+    - Check the `ruler-query-frontend` logs to find out more details about the error
+  - **High latency**
+    - Check the `Mimir / Remote ruler reads resources` dashboard to see if CPU or Memory usage increased unexpectedly
```
Comment (osg-grafana) [inline docs/sources/operators-guide/mimir-runbooks/_index.md:361]:
```suggestion
- Check the `Mimir / Remote ruler reads` dashboard to see if the issue is caused by failures or high latency:
  - **Failures**
    - Check the `ruler-query-frontend` logs to find out more details about the error.
  - **High latency**
    - Check the `Mimir / Remote ruler reads resources` dashboard to see if CPU or memory usage increased unexpectedly.
```

**Review by osg-grafana — APPROVED** [review]
Unblocking with feedback

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 5]
```diff
@@ -640,6 +640,25 @@
             ||| % $._config,
           },
         },
+        {
+          alert: $.alertName('RulerRemoteEvaluationFailing'),
+          expr: |||
+            100 * (
+            sum by (%s) (rate(cortex_request_duration_seconds_count{route="/httpgrpc.HTTP/Handle", status_code!~"5..", job=~".*/ruler-query-frontend"}[5m]))
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts.libsonnet:647]: Please don't hardcode `job=~".*/ruler-query-frontend"` but use `$.jobMatcher($._config.job_names.ruler_query_frontend)` instead.

Comment (ortuman) [reply operations/mimir-mixin/alerts/alerts.libsonnet]: Changed @ e91ae0aa22e7d95f7d1039aa523c65ae6ad512b2

File: operations/mimir-mixin/alerts/alerts-utils.libsonnet [hunk 6]
```diff
@@ -8,4 +8,11 @@
   // The alert name is prefixed with the product name (eg. AlertName -> MimirAlertName).
   alertName(name)::
     $._config.product + name,
+
+  // The mixin allow specialism of the job selector depending on if its a single binary
+  // deployment or a namespaced one.
+  jobMatcher(job)::
+    if $._config.singleBinary
+    then 'job=~"$job"'
+    else '%s=~"$cluster", job=~"($namespace)/(%s)"' % [$._config.per_cluster_label, job],
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts-utils.libsonnet:17]: This has been copy-pasted from the dashboard util, but in alerts we don't have the dashboard variables like `$job`, `$cluster` or `$namespace`. Here we want to match any namespace, so it's easier:


```suggestion
    'job=~".*/%s"' % job,
```

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 7]
```diff
@@ -640,6 +640,25 @@
             ||| % $._config,
           },
         },
+        {
+          alert: $.alertName('RulerRemoteEvaluationFailing'),
+          expr: |||
+            100 * (
+            sum by (%s) (rate(cortex_request_duration_seconds_count{route="/httpgrpc.HTTP/Handle", status_code!~"5..", %s}[5m]))
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts.libsonnet:647]: I guess you to track the rate of 5xx errors, right?
```suggestion
            sum by (%s) (rate(cortex_request_duration_seconds_count{route="/httpgrpc.HTTP/Handle", status_code=~"5..", %s}[5m]))
```

Comment (ortuman) [reply operations/mimir-mixin/alerts/alerts.libsonnet]: done @ 13a267c8baf9e1facfcbd699647163bc3d4a0bf1

**Review by pracucci — APPROVED** [review]
LGTM, thanks|

**Pull request merged**